core/trino-main/src/main/java/io/trino/operator/SpoolingPagePartitioner.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.spi.Page;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+
+import static com.google.common.base.Verify.verify;
+import static io.trino.execution.buffer.PageSplitterUtil.splitPage;
+import static java.lang.Math.clamp;
+
+public class SpoolingPagePartitioner
+{
+    static final double LOWER_BOUND = 0.05; // 5% of the target size
+    static final double UPPER_BOUND = 0.1; // 10% of the target size
+
+    private SpoolingPagePartitioner() {}
+
+    public static List<List<Page>> partition(List<Page> pages, long targetSize)
```

Comment:
Have you considered using `io.trino.execution.buffer.PageSplitterUtil#splitPage(io.trino.spi.Page, long)` instead ?